### PR TITLE
CI: Add linter job to CI workflow

### LINKED ISSUE #165: Add a go linter CI step
The CI should run a go linter step to avoid fixing linter issues in pull requests.

Once its enabled, there would inevitably be linting issues throughout the codebase. Fix them as part of the pull request for adding the linter in CI.

### PR DESCRIPTION
**Pull request by TheodosiouTh**
Closes #165

### CODE REVIEW

**Comment by preslavmihaylov** [comment]
@TheodosiouTh it seems there are some more linting issues once I ran it locally.

Note that the github action you've just added doesn't run on this PR's CI so we have to make sure the check passes locally.

Pasting the output:
image

**Comment by TheodosiouTh** [comment]
@preslavmihaylov Thank you for your input.
The pr is still a "work in progress" as I could not get it done yesterday.

**Comment by preslavmihaylov** [comment]
Also, could you tweak the linter config a bit and include the `golint` and `gofmt` linters which are disabled by default.

Additionally, once you make that change, you can remove the gofmt step from the CI pipeline as its a subset of the step you're adding.

**Comment by preslavmihaylov** [comment]
Actually, we should use `revive` as it seems `golint` is deprecated

**Comment by TheodosiouTh** [comment]
@preslavmihaylov  I added `revive` and  `gofmt`  as you requested. I also removed the `gofmt` step and fixed all the linting errors and warnings. 
[Here is the final run of the workflow](https://github.com/TheodosiouTh/todocheck/runs/3517172940). 
The tests fail but I don't know why (on this PR they pass). I will keep looking at it to see what the issue is.

**Comment by TheodosiouTh** [comment]
@preslavmihaylov  I have checked the code again and I do not get why the `make test` would fail after all the tests pass. My guess is that the error occurs because of the environment variables. So I think this will not be an issue when we merge the PR.

**Comment by TheodosiouTh** [comment]
@preslavmihaylov  I think the PR is ready for review, feel free to check it and give me feedback.

File: main.go [hunk 1]
```diff
@@ -77,7 +77,10 @@ func main() {
 	}
 
 	if len(todoErrs) > 0 {
-		printTodoErrs(todoErrs, *format)
+		err = printTodoErrs(todoErrs, *format)
+		if err != nil {
+			logger.Info("Error when printing todo errors" + err.Error())
```
Comment (preslavmihaylov) [inline main.go:82]: This should be a panic as we should never get here

File: testing/scenariobuilder/builder.go [hunk 2]
```diff
@@ -335,7 +336,10 @@ func (s *TodocheckScenario) setupMockIssueTrackerServer() (teardownFunc, error)
 
 		for issue := range s.issues {
 			if r.URL.Path == issuetracker.IssueURLFrom(s.issueTracker, issue) {
-				w.Write(issuetracker.BuildResponseFor(s.issueTracker, issue, s.issues[issue]))
+				_, err := w.Write(issuetracker.BuildResponseFor(s.issueTracker, issue, s.issues[issue]))
+				if err != nil {
+					logger.Info("Error " + err.Error() + " occured, when writing issuetracker response of issue:" + issue)
```
Comment (preslavmihaylov) [inline testing/scenariobuilder/builder.go:341]: This should be a panic as it should never happen

File: traverser/todoerrs/todoerrs.go [hunk 3]
```diff
@@ -32,7 +33,10 @@ func commentsCallback(chk *checker.Checker, customTodos []string, todoErrCallbac
 		if err != nil {
 			return fmt.Errorf("couldn't check todo line: %w", err)
 		} else if todoErr != nil {
-			todoErrCallback(todoErr)
+			err = todoErrCallback(todoErr)
+			if err != nil {
+				logger.Info("couldn't run todo error callback: " + err.Error())
```
Comment (preslavmihaylov) [inline traverser/todoerrs/todoerrs.go:38]: We should instead return `fmt.Errorf("received error from todo err callback: %w", err)`

File: traverser/comments/comments.go [hunk 4]
```diff
@@ -42,14 +43,19 @@ func (t *Traverser) TraversePath(path string) error {
...
+				logger.Info("Error Occured during handleStateChange of file:" + filename + " at line: " + line)
+			}
 			prev = curr
 		}
 
 		curr = next
 		next = 0
-		t.handleStateChange(filename, line, linecnt, prev, curr, next)
+		err := t.handleStateChange(filename, line, linecnt, prev, curr, next)
+		if err != nil {
```
Comment (preslavmihaylov) [inline traverser/comments/comments.go:56]: These errors shouldn't be handled here as they are handled via the last `return t.callbackErr`.

Does the linter allow ignoring the error via `_ = t.handleStateChange(...)`?

Alternatively, you can change the signature of the function to not return an `error` and update it accordingly

**Review by preslavmihaylov — CHANGES_REQUESTED** [review]

File: main.go (hunk 1 above)
Comment (TheodosiouTh) [reply main.go]: I changed the code to panic if this point is reached during execution.

File: testing/scenariobuilder/builder.go (hunk 2 above)
Comment (TheodosiouTh) [reply testing/scenariobuilder/builder.go]: I changed the code to panic if this point is reached during execution.

File: traverser/todoerrs/todoerrs.go (hunk 3 above)
Comment (TheodosiouTh) [reply traverser/todoerrs/todoerrs.go]: I updated the code according to your suggestion.

File: traverser/comments/comments.go (hunk 4 above)
Comment (TheodosiouTh) [reply traverser/comments/comments.go:56]: The linter does allow `_ = t.handleStageChange(...)` so I updated the code.

**Comment by TheodosiouTh** [comment]
@preslavmihaylov I updated the code according to your suggestions. Also, I found out a logic bug in `printTodoErrs`, it was always returning an error even if an accepted format was provided (caused missing early return statements). I added the early return statements and changed the `if- else if` to a `switch-case` as I find switch case easier to read. You can see the bug fix on this (commit)[https://github.com/preslavmihaylov/todocheck/pull/166/commits/f0aac5aa40cb6262828b55cea10666d4e23dba88].

Also feel free to check the rest of the changes.

**Review by preslavmihaylov — APPROVED** [review]
Thank you for your contribution!

**Pull request merged**